src/expr/src/relation/canonicalize.rs

```diff
@@ -83,6 +87,85 @@ pub fn canonicalize_predicates(predicates: &mut Vec<MirScalarExpr>, input_type:
             predicates.push(expr);
         }
     }
+
+    // 3) Reduce across `predicates`.
+    // If a predicate `p` cannot be null, and `f(p)` is of type bool
+    // (`f(p)` is allowed to be nullable), then the predicate `p & f(p)` is
+    // equal to `p & f(true)`, and `!p & f(p)` is equal to `!p & f(false)`.
+    // For any index i, the `Vec` of predicates `[p1, ... pi, ... pn]` is
+    // equivalent to the single predicate `pi & (p1 & ... & p(i-1) & p(i+1)
+    // ... & pn)`.
+    // Thus, if `pi` (resp. `!pi`) cannot be null, it is valid to replace with
+    // `true` (resp. `false`) every subexpression in
+    // `(p1 & ... & p(i-1) & p(i+1) ... & pn)` that is equal to `pi`.
+
+    // Note that this also dedups predicates since if `p1 = p2` then this
+    // reduction process will replace `p1` with true.
+
+    // Pop out a predicate `p` from `predicates_to_apply`. Check all the other
+    // predicates (this is the union of `predicates_to_apply` and
+    // `applied_predicates`), and if `p` is a subexpression of any of those
+    // predicates, replace the subexpression with `true`. Then stick `p` into
+    // `applied_predicates`. If a predicate `q` in `applied_predicates` was
+    // altered in the aforementioned process, remove `q` from
+    // `applied_predicates` and put it in `predicates_to_apply` so we can reduce
+    // the predicates even further.
+    let mut predicates_to_apply = Vec::new();
+    let mut applied_predicates = Vec::<MirScalarExpr>::new();
+    // Seed `predicates_to_apply` with the `Vec` of all predicates.
+    std::mem::swap(&mut predicates_to_apply, predicates);
+
+    while let Some(predicate_to_apply) = predicates_to_apply.pop() {
+        let mut replace_subexpr_other_predicates =
+            |expr: &MirScalarExpr, constant_bool: &MirScalarExpr| {
+                // Do not replace subexpressions equal to `expr` if `expr` is a
+                // literal to avoid infinite looping.
+                if !expr.is_literal() && !expr.typ(input_type).nullable {
```

Comment:
I am not sure if the code has the concept of a ```deterministic``` function but I think non-deterministic expressions should also be excluded from the optimization.

Right now there seems to be only one such function, ```mz_sleep``` and expressions containing it are reduced aggressively:

```
materialize=> EXPLAIN SELECT * FROM t1 WHERE mz_internal.mz_sleep(col_not_null) > mz_internal.mz_sleep(col_not_null) AND (mz_internal.mz_sleep(col_not_null) > mz_internal.mz_sleep(col_not_null) = true);
                       Optimized Plan                       
------------------------------------------------------------
 %0 =                                                      +
 | Get materialize.public.t1 (u227)                        +
 | Filter (mz_sleep(i32tof64(#1)) > mz_sleep(i32tof64(#1)))+
```

So the 4 sleeps have been reduced to 2. Without delving into @frankmcsherry's question as to whether in SQL one is required to evaluate the entire expression or only parts thereof, it seems safer to simply exclude non-deterministic expressions from any such optimizations/rewrites.